src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt

```diff
@@ -0,0 +1,191 @@
+    }
+
+    return offenderTransformer.transform(prisoner)
+  }
+
+  fun getNextPrisonerIdentifier(): PrisonerIdentifier {
+    var retries = 0
+    var updated: Boolean
+    var nextSequence: NomsIdSequence
+    var currentSequence: NomsIdSequence
+    do {
+      currentSequence = prisonerRepository.nomsIdSequence
+      nextSequence = currentSequence.next()
+      updated = prisonerRepository.updateNomsIdSequence(nextSequence, currentSequence) > 0
+    } while (!updated && retries++ < 10)
+    if (!updated) {
+      throw RuntimeException("Prisoner Identifier cannot be generated, please try again")
+    }
+    return PrisonerIdentifier().withId(currentSequence.prisonerIdentifier)
+  }
+
+  private fun gender(code: String): Result<Gender> =
+    genderRepository.findByIdOrNull(ReferenceCode.Pk(Gender.SEX, code))?.let { success(it) } ?: failure(
+      EntityNotFoundException.withMessage("Gender $code not found")
+    )
+
+  private fun ethnicity(code: String?): Result<Ethnicity>? =
+    code?.takeIf { it.isNotBlank() }?.let {
+      ethnicityRepository.findByIdOrNull(ReferenceCode.Pk(Ethnicity.ETHNICITY, code))?.let { success(it) }
+        ?: failure(EntityNotFoundException.withMessage("Ethnicity $code not found"))
+    }
+
+  private fun title(code: String?): Result<Title>? =
+    code?.takeIf { it.isNotBlank() }?.let {
+      titleRepository.findByIdOrNull(ReferenceCode.Pk(Title.TITLE, code))?.let { success(it) }
+        ?: failure(EntityNotFoundException.withMessage("Title $code not found"))
+    }
+
+  private fun suffix(code: String?): Result<Suffix>? =
+    code?.takeIf { it.isNotBlank() }?.let {
+      suffixRepository.findByIdOrNull(ReferenceCode.Pk(Suffix.SUFFIX, code))?.let { success(it) }
+        ?: failure(EntityNotFoundException.withMessage("Suffix $code not found"))
+    }
+
+  private fun validPncNumber(pncNumber: String?): Result<String>? =
+    pncNumber?.takeIf { it.isNotBlank() }?.let { pnc ->
+      return Pnc.getShortPncNumber(pnc).checkForDuplicatePncNumber().map {
+        return Pnc.getLongPncNumber(pnc).checkForDuplicatePncNumber()
+      }
+    }
+
+  private fun String.checkForDuplicatePncNumber(): Result<String> =
+    offenderIdentifierRepository.findByIdentifierTypeAndIdentifier("PNC", this)
+      .takeIf { matches -> matches.isNotEmpty() }?.let { matches ->
+        failure(matches.toPncMatchFailure())
+      } ?: success(Pnc.getLongPncNumber(this))
+
+  private fun validCroNumber(croNumber: String?): Result<String>? =
+    croNumber?.takeIf { it.isNotBlank() }?.checkForDuplicateCroNumber()
+
+  private fun String.checkForDuplicateCroNumber(): Result<String> =
+    offenderIdentifierRepository.findByIdentifierTypeAndIdentifier("CRO", this)
+      .takeIf { matches -> matches.isNotEmpty() }?.let { matches ->
+        failure(matches.toCroMatchFailure())
+      } ?: success(this)
+
+  private fun validUniqueUppercaseNames(requestToCreate: RequestToCreate): Result<Pair<String, String>> {
```

Comment:
Minor, but all these `Pair<String, String>` could be a value class for readability. Maybe called `FirstAndLastName`.